File: frontend/src/types/drone.ts
// available and in_flight follow the missions this drone is booked on,
// the other two are the admin taking it out of rotation by hand
export type DroneStatus = 'available' | 'in_flight' | 'maintenance' | 'retired'

export interface Drone {
  id: string
  name: string
  model: string
  serial: string
  // where the aircraft puts its video, the jetson serves rtsp on the airframe
  stream_url: string
  status: DroneStatus
  owner_id: string
  flight_hours: number
  missions_flown: number
  // the mission holding this aircraft, a booking is exclusive until it is flown
  booked_on: string | null
  // removed from the fleet but kept, the missions it flew still read from it
  hidden: boolean
  created_at: string
  updated_at: string
}

export interface DroneCreateInput {
  name: string
  model?: string
  serial?: string
  stream_url?: string
  status?: DroneStatus
}

export interface DroneUpdateInput {
  name?: string
  model?: string
  serial?: string
  stream_url?: string
  status?: DroneStatus
}

File: backend/app/models/drone.py
from datetime import datetime, timezone
from enum import Enum

from beanie import Document
from pydantic import Field


class DroneStatus(str, Enum):
    # available and in_flight move on their own as missions are flown, the other
    # two are the admin taking an aircraft out of rotation by hand
    AVAILABLE = "available"
    IN_FLIGHT = "in_flight"
    MAINTENANCE = "maintenance"
    RETIRED = "retired"


class Drone(Document):
    name: str
    model: str = ""
    serial: str = ""
    # where the aircraft puts its video, the jetson serves rtsp on the airframe
    stream_url: str = ""
    # retired is how an aircraft leaves the fleet, the record stays so the
    # missions it flew can still say what flew them
    status: DroneStatus = DroneStatus.AVAILABLE
    owner_id: str  # the admin whose fleet this belongs to
    # the digital twin's running totals, both fed by submitted pilot reports
    flight_hours: float = 0
    missions_flown: int = 0
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "drones"
